dbms/src/Interpreters/CollectJoinOnKeysVisitor.cpp

```diff
@@ -32,16 +32,24 @@ void CollectJoinOnKeysMatcher::Data::addJoinKeys(const ASTPtr & left_ast, const
 }
 
 void CollectJoinOnKeysMatcher::Data::addAsofJoinKeys(const ASTPtr & left_ast, const ASTPtr & right_ast,
-                                                     const std::pair<size_t, size_t> & table_no)
+                                                     const std::pair<size_t, size_t> & table_no, const ASOF::Inequality & inequality)
 {
     if (table_no.first == 1 || table_no.second == 2)
     {
         asof_left_key = left_ast->clone();
         asof_right_key = right_ast->clone();
+        analyzed_join.setAsofInequality(inequality);
+        return;
+    }
+    else if (table_no.first == 2 || table_no.second == 1)
+    {
+        asof_left_key = right_ast->clone();
+        asof_right_key = left_ast->clone();
+        analyzed_join.setAsofInequality(ASOF::reverseInequality(inequality));
         return;
     }
 
-    throw Exception("ASOF JOIN for (left_table.x <= right_table.x) is not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    throw Exception("ASOF JOIN requires keys inequality from different tables", ErrorCodes::NOT_IMPLEMENTED);
```

Comment:
Does this case ever happen? For same table on both sides, I get a different error:
```
:) SELECT * FROM A ASOF INNER JOIN B ON A.t <= A.t AND A.a == B.b;

-- Invalid expression for JOIN ON. In expression t <= t columns t and t are from the same table but from different arguments of equal function.
```
Maybe we should switch just on table_no.first, and check the other one with assert().